Clean up AsyncMKM's shutdown_all

vidartf opened the pull request:
This enables the manager to be reused after shutdown_all by:
- Making sure all the cleanup that happens in shutdown_kernel gets called.
- Making sure to await any starting kernels, and also shut them down

Partially addresses #616.

vidartf commented:
Hmm, seems as if all the 3.6 tests hang here :-/

vidartf commented:
So now mac 3.6 is the only one that hangs? What is even going on 😄

kevin-bates commented:
FWIW, I just pulled your branch and enabled CI on macOS 3.8 and 3.9 (to see what the issues were) and got hangs on `ubuntu` and `windows` (3.6 for both), but all four macOS runs completed successfully: https://github.com/kevin-bates/jupyter_client/actions/runs/556331259

kevin-bates reviewed (commented):
The changes good Vidar - thanks.

I'm seeing this in the test output...
```
/Users/runner/work/jupyter_client/jupyter_client/jupyter_client/tests/test_multikernelmanager.py:188: RuntimeWarning: coroutine 'AsyncKernelManager.is_alive' was never awaited
    assert km.is_alive(kid)
```
Could you please add an `await` on that line?

vidartf commented:
> Could you please add an `await` on that line?

Good spot, seems like that fixed the hanging as well?

kevin-bates commented:
> seems like that fixed the hanging as well

I wish.  I just re-ran the jobs and it's hung for all three py3.6 runs.  This is the same kind of behavior I found while looking into this yesterday.  The hang appears to be completely related to 3.6 but intermittent as to which platform exhibits the hang (and sometimes succeeding).

kevin-bates commented on jupyter_client/tests/test_multikernelmanager.py:
```diff
@@ -230,6 +230,39 @@ async def test_shutdown_all(self):
+    @gen_test(timeout=20)
+    async def test_shutdown_all_while_starting(self):
+        km = self._get_tcp_km()
+        kid_future = km.start_kernel(stdout=PIPE, stderr=PIPE)
```
I believe we should wrap this result in an `ensure_future()`.  After doing so, I was able to run the tests on 3.6 successfully multiple times w/o failures.
```suggestion
        kid_future = asyncio.ensure_future(km.start_kernel(stdout=PIPE, stderr=PIPE))
```

kevin-bates commented:
Hi @vidartf - all signs are pointing at the update I referenced in my last comment - at least the tests haven't failed since that update (with about 6 runs against the 3.6 platforms).

I am seeing this warning in the output, mostly on Windows but also on Mac but not consistently.  
```
upyter_client/tests/test_multikernelmanager.py::TestAsyncKernelManager::test_shutdown_all_while_starting PASSED
c:\hostedtoolcache\windows\python\3.6.8\x64\lib\multiprocessing\process.py:93: RuntimeWarning: coroutine 'TestAsyncKernelManager.raw_tcp_lifecycle' was never awaited
  self._target(*self._args, **self._kwargs)
jupyter_client/tests/test_multikernelmanager.py::TestAsyncKernelManager::test_start_parallel_process_kernels PASSED
```
It kinda makes sense since the new process doesn't have an ioloop to deal with async targets.  I haven't looked into this aspect of things very deeply but wondering if this would need something like a ProcessPoolExecutor along with a run-in-executor call.  At any rate, something for another day.

vidartf commented:
@kevin-bates  I fixed the warning you pointed out. I also fixed a deprecation warning for random sampling from a set (sorry for being lazy and just adding it onto this PR instead of creating a new one 😉).

vidartf commented:
Haha, I fixed it for the spawn multiprocess, but broke the fork one, looking.

vidartf commented:
I have to admit I don't fully understand what happens with event loops when forking (I don't fully understand the details of forking either), so probably someone should pay some extra attention to that bit.

kevin-bates commented:
> I have to admit I don't fully understand what happens with event loops when forking

I'm afraid I'm in the same boat and, of course, [the behaviors differ by platform](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods).

kevin-bates reviewed (approved):
Thanks Vidar!
